cpu/stm32_common/periph/pm.c

```diff
@@ -41,6 +48,176 @@
+    }
+
+    // lpm_portmask_system[((uint32_t)&port>> 10) & 0x0f] |= 1 << pin;
+    lpm_portmask_system[_port_num(pin)] |= 1 << pin_num;
+}
+
+/* Do not change GPIO state in sleep mode */
+void lpm_arch_add_gpio_exclusion(gpio_t gpio) {
+	uint8_t port = ((uint32_t)gpio >> 10) & 0x0f;
+	uint8_t pin = ((uint32_t)gpio & 0x0f);
+
+	lpm_portmask_user[port] |= (uint16_t)(1<<pin);
+}
+
+/* Change GPIO state to AIN in sleep mode */
+void lpm_arch_del_gpio_exclusion(gpio_t gpio) {
+	uint8_t port = ((uint32_t)gpio >> 10) & 0x0f;
+	uint8_t pin = ((uint32_t)gpio & 0x0f);
+
+	lpm_portmask_user[port] &= ~(uint16_t)(1<<pin);
+}
+
+void lpm_before_i_go_to_sleep(void){
+    uint8_t i;
+    uint8_t p;
+    uint32_t mask;
+    GPIO_TypeDef *port;
+
+
+    for (i = 0; i < CPU_NUMBER_OF_PORTS; i++) {
+        port = (GPIO_TypeDef *)(GPIOA_BASE + i*(GPIOB_BASE - GPIOA_BASE));
+
+        /* save GPIO registers values */
+        lpm_gpio_moder[i] = port->MODER;
+        lpm_gpio_pupdr[i] = port->PUPDR;
+        lpm_gpio_otyper[i] = (uint16_t)(port->OTYPER & 0xFFFF);
+        lpm_gpio_odr[i] = (uint16_t)(port->ODR & 0xFFFF);
+	}
+
+    /* specifically set GPIOs used for external SPI devices */
+    if (SPI_0_ISON()) {
+        pin_set(spi_config[0].sclk_pin, 0);
+        pin_set(spi_config[0].mosi_pin, 0);
+        pin_set(spi_config[0].cs_pin, 1);
+    } else {
+		p = ((uint32_t)PORT_A >> 10) & 0x0f;
+		lpm_portmask_system[p] &= ~(1 << 5);
+		lpm_portmask_system[p] &= ~(1 << 7);
+        p = ((uint32_t)PORT_B >> 10) & 0x0f;
+		lpm_portmask_system[p] &= ~(1 << 6);
+	}
+
+    /* save GPIO clock configuration */
+    ahb_gpio_clocks = RCC->AHBENR & 0xFF;
+    /* enable all GPIO clocks */
+    periph_clk_en(AHB, 0xFF);
+
+    for (i = 0; i < CPU_NUMBER_OF_PORTS; i++) {
+           port = (GPIO_TypeDef *)(GPIOA_BASE + i*(GPIOB_BASE - GPIOA_BASE));
+           mask = 0xFFFFFFFF;
+           if (EXTI->IMR) {
+               for (p = 0; p < 16; p ++) {
+                   /* exclude GPIOs we previously set with pin_set */
+                   if ((lpm_portmask_system[i] | lpm_portmask_user[i]) & (1 << p)) {
+                       mask &= ~(0x03 << (p*2));
+                   } else {
+                       if ((EXTI->IMR & (1 << p)) && ((((SYSCFG->EXTICR[p >> 2]) >> ((p & 0x03) * 4)) & 0xF) == i)) {
+                           mask &= ~(0x03 << (p*2));
+                       }
+                   }
+               }
+           }
+
+           /* disable pull-ups on GPIOs */
+           tmpreg = port->PUPDR;
+           tmpreg &= ~mask;
+           port->PUPDR = tmpreg;
+
+           /* set GPIOs to AIN mode */
+           tmpreg = port->MODER;
+           tmpreg |= mask;
+           port->MODER = tmpreg;
+       }
+
+    /* restore GPIO clocks */
+    tmpreg = RCC->AHBENR;
+    tmpreg &= ~((uint32_t)0xFF);
+    tmpreg |= ahb_gpio_clocks;
+    periph_clk_en(AHB, tmpreg);
+}
+
+/* restore GPIO settings */
+static void lpm_when_i_wake_up (void) {
```

Comment:
Same here as in the "prepare for sleep" function. Better to configure each peripheral separately and not all at once.